File: src/python/xerxes/tools/home_assistant_tools.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Home Assistant integration tools for controlling smart home devices and querying entity states.

This module provides tools for interacting with Home Assistant instances, including
listing entities, getting states, and calling services.

Example:
    >>> from xerxes.tools.home_assistant_tools import HomeAssistantClient, ha_list_entities
    >>> entities = ha_list_entities.static_call(domain="light")
"""

from __future__ import annotations

import json
import logging
import os
import threading
import typing as tp

from ..types import AgentBaseFn

logger = logging.getLogger(__name__)


class HomeAssistantClient:
    """Client for interacting with Home Assistant REST API.

    This class provides a thread-safe interface to Home Assistant's API, enabling
    agents to query entity states, list services, and control smart home devices.

    Uses environment variables HASS_BASE_URL and HASS_TOKEN for configuration,
    or can be instantiated with explicit parameters.

    Attributes:
        _instance: Singleton instance for the client.
        _lock: Thread synchronization lock.

    Example:
        >>> client = HomeAssistantClient.instance()
        >>> states = client.list_states()
        >>> client.call_service("light", "turn_on", {"entity_id": "light.living_room"})
    """

    _instance: HomeAssistantClient | None = None
    _lock = threading.Lock()

    def __init__(
        self,
        base_url: str | None = None,
        token: str | None = None,
        http_client: tp.Any | None = None,
    ) -> None:
        """Initialize the Home Assistant client.

        Args:
            base_url: Home Assistant URL. Defaults to HASS_BASE_URL env var.
            token: Long-lived access token. Defaults to HASS_TOKEN env var.
            http_client: Optional custom HTTP client for testing.
        """
        self.base_url = (base_url or os.environ.get("HASS_BASE_URL", "")).rstrip("/")
        self.token = token or os.environ.get("HASS_TOKEN", "")
        self._http = http_client

    @classmethod
    def instance(cls) -> HomeAssistantClient:
        """Get the singleton client instance.

        Returns:
            The shared HomeAssistantClient instance.
        """
        with cls._lock:
            if cls._instance is None:
                cls._instance = cls()
            return cls._instance

    @classmethod
    def install_for_test(
        cls,
        *,
        base_url: str = "http://hass.test",
        token: str = "tok",
        http_client: tp.Any,
    ) -> HomeAssistantClient:
        """Install a test client for unit testing.

        Args:
            base_url: Test server URL.
            token: Test authentication token.
            http_client: Mock HTTP client.

        Returns:
            The configured test client instance.
        """
        with cls._lock:
            cls._instance = cls(base_url=base_url, token=token, http_client=http_client)
            return cls._instance

    @classmethod
    def reset(cls) -> None:
        """Reset the singleton instance to None."""
        with cls._lock:
            cls._instance = None

    def list_states(self) -> list[dict[str, tp.Any]]:
        """Get all entity states from Home Assistant.

        Returns:
            List of entity state dictionaries.
        """
        return self._get("/api/states")

    def get_state(self, entity_id: str) -> dict[str, tp.Any] | None:
        """Get the state of a specific entity.

        Args:
            entity_id: Full entity ID (e.g., "light.living_room").

        Returns:
            Entity state dictionary, or None if not found.
        """
        try:
            return self._get(f"/api/states/{entity_id}")
        except Exception:
            return None

    def list_services(self) -> list[dict[str, tp.Any]]:
        """Get available services from Home Assistant.

        Returns:
            List of service domain dictionaries.
        """
        return self._get("/api/services")

    def call_service(
        self,
        domain: str,
        service: str,
        data: dict[str, tp.Any] | None = None,
    ) -> list[dict[str, tp.Any]]:
        """Call a Home Assistant service.

        Args:
            domain: Service domain (e.g., "light", "switch").
            service: Service name (e.g., "turn_on", "toggle").
            data: Optional service data dictionary.

        Returns:
            List of service call results.
        """
        return self._post(f"/api/services/{domain}/{service}", json_body=data or {})

    def _headers(self) -> dict[str, str]:
        """Build HTTP headers for API requests."""
        h = {"Content-Type": "application/json"}
        if self.token:
            h["Authorization"] = f"Bearer {self.token}"
        return h

    def _url(self, path: str) -> str:
        """Build full URL from path.

        Args:
            path: API endpoint path.

        Returns:
            Full URL string.
        """
        if not self.base_url:
            raise RuntimeError("HomeAssistantClient: HASS_BASE_URL is not configured")
        return f"{self.base_url}{path}"

    def _get(self, path: str) -> tp.Any:
        """Perform GET request to Home Assistant API.

        Args:
            path: API endpoint path.

        Returns:
            Parsed JSON response.
        """
        url = self._url(path)
        if self._http is not None:
            resp = self._http.get(url, headers=self._headers())
            return _parse(resp)
        try:
            import httpx
        except ImportError as exc:
            raise RuntimeError("httpx required for HomeAssistantClient") from exc
        resp = httpx.get(url, headers=self._headers(), timeout=15.0)
        resp.raise_for_status()
        return resp.json()

    def _post(self, path: str, json_body: dict[str, tp.Any]) -> tp.Any:
        """Perform POST request to Home Assistant API.

        Args:
            path: API endpoint path.
            json_body: JSON body for the request.

        Returns:
            Parsed JSON response.
        """
        url = self._url(path)
        if self._http is not None:
            resp = self._http.post(url, json=json_body, headers=self._headers())
            return _parse(resp)
        try:
            import httpx
        except ImportError as exc:
            raise RuntimeError("httpx required for HomeAssistantClient") from exc
        resp = httpx.post(url, json=json_body, headers=self._headers(), timeout=15.0)
        resp.raise_for_status()
        return resp.json()


def _parse(resp: tp.Any) -> tp.Any:
    """Parse response to JSON.

    Args:
        resp: HTTP response object.

    Returns:
        Parsed JSON or raw body.
    """
    if hasattr(resp, "json") and callable(resp.json):
        try:
            return resp.json()
        except Exception:
            pass
    body = getattr(resp, "text", None) or getattr(resp, "body", None) or ""
    if isinstance(body, bytes):
        body = body.decode()
    try:
        return json.loads(body)
    except Exception:
        return body


def _filter_entities(
    states: list[dict[str, tp.Any]],
    *,
    domain: str | None = None,
    area: str | None = None,
) -> list[dict[str, tp.Any]]:
    """Filter entity states by domain or area.

    Args:
        states: List of entity states.
        domain: Filter by domain (e.g., "light").
        area: Filter by area ID.

    Returns:
        Filtered list of states.
    """
    out = []
    for s in states or []:
        eid = s.get("entity_id", "")
        if domain and not eid.startswith(f"{domain}."):
            continue
        if area:
            attrs = s.get("attributes") or {}
            entity_area = attrs.get("area_id") or attrs.get("area")
            if entity_area != area:
                continue
        out.append(s)
    return out


class ha_list_entities(AgentBaseFn):
    """List Home Assistant entities with optional filtering.

    Retrieves all or filtered entity states from Home Assistant.

    Example:
        >>> ha_list_entities.static_call(domain="light")
        >>> ha_list_entities.static_call(area="Living Room", limit=50)
    """

    @staticmethod
    def static_call(
        domain: str | None = None,
        area: str | None = None,
        limit: int = 200,
        **context_variables: tp.Any,
    ) -> dict[str, tp.Any]:
        """List Home Assistant entities.

        Args:
            domain: Filter by entity domain (e.g., "light", "switch").
            area: Filter by area name.
            limit: Maximum number of entities to return. Defaults to 200.
            **context_variables: Additional context passed through to downstream calls.

        Returns:
            Dictionary with 'count' and 'entities' list.
        """
        try:
            limit_n = int(limit) if limit is not None else 200
        except (TypeError, ValueError):
            limit_n = 200
        states = HomeAssistantClient.instance().list_states()
        filtered = _filter_entities(states or [], domain=domain, area=area)[:limit_n]
        return {
            "count": len(filtered),
            "entities": [
                {
                    "entity_id": e.get("entity_id"),
                    "state": e.get("state"),
                    "attributes": e.get("attributes", {}),
                }
                for e in filtered
            ],
        }


class ha_list_services(AgentBaseFn):
    """List available Home Assistant services.

    Retrieves all service domains or filters by specific domain.

    Example:
        >>> ha_list_services.static_call()
        >>> ha_list_services.static_call(domain="light")
    """

    @staticmethod
    def static_call(
        domain: str | None = None,
        **context_variables: tp.Any,
    ) -> dict[str, tp.Any]:
        """List Home Assistant services.

        Args:
            domain: Filter by service domain.
            **context_variables: Additional context passed through to downstream calls.

        Returns:
            Dictionary with 'domains' list.
        """
        catalog = HomeAssistantClient.instance().list_services() or []
        if domain:
            catalog = [d for d in catalog if d.get("domain") == domain]
        return {"domains": catalog}


class ha_get_state(AgentBaseFn):
    """Get the current state of a Home Assistant entity.

    Example:
        >>> ha_get_state.static_call(entity_id="light.living_room")
    """

    @staticmethod
    def static_call(
        entity_id: str,
        **context_variables: tp.Any,
    ) -> dict[str, tp.Any]:
        """Get entity state.

        Args:
            entity_id: Full entity ID to query.
            **context_variables: Additional context passed through to downstream calls.

        Returns:
            Dictionary with entity_id, state, attributes, and timestamps.
        """
        state = HomeAssistantClient.instance().get_state(entity_id)
        if not state or not isinstance(state, dict) or not state.get("entity_id"):
            return {"error": "not_found", "entity_id": entity_id}
        return {
            "entity_id": state.get("entity_id"),
            "state": state.get("state"),
            "attributes": state.get("attributes", {}),
            "last_changed": state.get("last_changed", ""),
            "last_updated": state.get("last_updated", ""),
        }


class ha_call_service(AgentBaseFn):
    """Call a Home Assistant service to control devices.

    Example:
        >>> ha_call_service.static_call(
        ...     domain="light",
        ...     service="turn_on",
        ...     data={"entity_id": "light.living_room"}
        ... )
    """

    @staticmethod
    def static_call(
        domain: str,
        service: str,
        data: dict[str, tp.Any] | None = None,
        **context_variables: tp.Any,
    ) -> dict[str, tp.Any]:
        """Call a Home Assistant service.

        Args:
            domain: Service domain (e.g., "light", "switch", "automation").
            service: Service to call (e.g., "turn_on", "toggle").
            data: Optional service data dictionary with entity_id and other parameters.
            **context_variables: Additional context passed through to downstream calls.

        Returns:
            Dictionary with 'ok' status and service response.
        """
        try:
            changed = HomeAssistantClient.instance().call_service(domain, service, data or {})
            return {"ok": True, "domain": domain, "service": service, "changed": changed}
        except Exception as exc:
            return {"ok": False, "error": str(exc), "domain": domain, "service": service}


__all__ = [
    "HomeAssistantClient",
    "ha_call_service",
    "ha_get_state",
    "ha_list_entities",
    "ha_list_services",
]

File: xerxes/src/tools/homeAssistantTools.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { ClientError, ConfigurationError, ValidationError } from '../core/errors.js'
import { ToolRegistry, type ToolExecutionContext } from '../executors/toolRegistry.js'
import { isJsonObject, type JsonObject, type JsonValue, type ToolDefinition } from '../types/toolCalls.js'
import { optionalInteger, optionalString, requiredString } from './inputs.js'

const DEFAULT_ENTITY_LIMIT = 200
const DEFAULT_TIMEOUT_MS = 15_000
const MAX_ENTITY_LIMIT = 1_000
const MAX_TIMEOUT_MS = 300_000
const DOMAIN_PATTERN = /^[a-z0-9_]+$/
const ENTITY_ID_PATTERN = /^[a-z0-9_]+(?:\.[a-z0-9_]+)+$/

/** Injectable native-fetch boundary for a deliberately configured Home Assistant instance. */
export type HomeAssistantFetch = (
  input: RequestInfo | URL,
  init?: RequestInit,
) => Promise<Response>

/**
 * Explicit Home Assistant REST configuration.
 *
 * Credentials are never read from process environment variables. Hosts must
 * provide a token directly, or deliberately opt into an unauthenticated local
 * gateway with allowUnauthenticated.
 */
export interface HomeAssistantClientOptions {
  readonly allowUnauthenticated?: boolean
  readonly baseUrl: string
  readonly fetchImplementation?: HomeAssistantFetch
  readonly timeoutMs?: number
  readonly token?: string
}

/** Safe typed error that never includes response bodies or credential values. */
export class HomeAssistantRequestError extends ClientError {
  readonly status: number | undefined

  constructor(message: string, status: number | undefined = undefined, cause: unknown = undefined) {
    super('home_assistant', message, cause, status === undefined ? {} : { status })
    this.status = status
  }
}

/**
 * Small REST client for a single configured Home Assistant installation.
 *
 * This intentionally has no singleton and no environment fallback. A host can
 * instantiate one client per tenant/session and inject a mock or hardened fetch
 * implementation at its network boundary.
 */
export class HomeAssistantClient {
  private readonly allowUnauthenticated: boolean
  private readonly baseUrl: URL
  private readonly fetchImplementation: HomeAssistantFetch
  private readonly timeoutMs: number
  private readonly token: string

  constructor(options: HomeAssistantClientOptions) {
    this.allowUnauthenticated = options.allowUnauthenticated ?? false
    this.baseUrl = normalizeBaseUrl(options.baseUrl)
    this.fetchImplementation = options.fetchImplementation ?? fetch
    this.timeoutMs = boundedTimeout(options.timeoutMs ?? DEFAULT_TIMEOUT_MS)
    this.token = normalizedToken(options.token)

    if (!this.token && !this.allowUnauthenticated) {
      throw new ConfigurationError(
        'homeAssistant.token',
        'must be supplied explicitly; set allowUnauthenticated only for an intentionally unauthenticated gateway',
      )
    }
  }

  async listStates(signal?: AbortSignal): Promise<readonly JsonObject[]> {
    return requiredObjectArray(await this.requestJson('GET', 'api/states', undefined, signal), 'states')
  }

  async getState(entityId: string, signal?: AbortSignal): Promise<JsonObject | undefined> {
    const normalizedEntityId = homeAssistantEntityId(entityId)
    try {
      const response = await this.requestJson(
        'GET',
        'api/states/' + encodeURIComponent(normalizedEntityId),
        undefined,
        signal,
      )
      if (response === undefined) {
        throw new HomeAssistantRequestError('Home Assistant returned an empty state response')
      }
      return requiredObject(response, 'state')
    } catch (error) {
      if (error instanceof HomeAssistantRequestError && error.status === 404) {
        return undefined
      }
      throw error
    }
  }

  async listServices(signal?: AbortSignal): Promise<readonly JsonObject[]> {
    return requiredObjectArray(await this.requestJson('GET', 'api/services', undefined, signal), 'services')
  }

  async callService(
    domain: string,
    service: string,
    data: JsonObject = {},
    signal?: AbortSignal,
  ): Promise<readonly JsonObject[]> {
    const normalizedDomain = homeAssistantDomain(domain, 'domain')
    const normalizedService = homeAssistantDomain(service, 'service')
    const response = await this.requestJson(
      'POST',
      'api/services/' + encodeURIComponent(normalizedDomain) + '/' + encodeURIComponent(normalizedService),
      data,
      signal,
    )
    if (response === undefined) {
      return []
    }
    return requiredObjectArray(response, 'service response')
  }

  private async requestJson(
    method: 'GET' | 'POST',
    path: string,
    body: JsonObject | undefined,
    signal: AbortSignal | undefined,
  ): Promise<JsonValue | undefined> {
    const headers: Record<string, string> = { Accept: 'application/json' }
    if (body !== undefined) {
      headers['Content-Type'] = 'application/json'
    }
    if (this.token) {
      headers.Authorization = 'Bearer ' + this.token
    }

    const init: RequestInit = {
      headers,
      method,
      signal: requestSignal(signal, this.timeoutMs),
    }
    if (body !== undefined) {
      init.body = JSON.stringify(body)
    }

    let response: Response
    try {
      response = await this.fetchImplementation(this.endpoint(path), init)
    } catch (error) {
      throw new HomeAssistantRequestError('could not reach the configured Home Assistant instance', undefined, error)
    }
    if (!response.ok) {
      throw new HomeAssistantRequestError('Home Assistant request failed with HTTP ' + response.status, response.status)
    }

    let text: string
    try {
      text = await response.text()
    } catch (error) {
      throw new HomeAssistantRequestError('could not read the Home Assistant response', response.status, error)
    }
    if (!text.trim()) {
      return undefined
    }

    try {
      const parsed: unknown = JSON.parse(text)
      if (!isJsonValue(parsed)) {
        throw new Error('response was not JSON-serializable')
      }
      return parsed
    } catch (error) {
      throw new HomeAssistantRequestError('Home Assistant returned invalid JSON', response.status, error)
    }
  }

  private endpoint(path: string): URL {
    const normalized = path.replace(/^\/+/, '')
    if (!normalized || normalized.includes('?') || normalized.includes('#')) {
      throw new ValidationError('homeAssistant.path', 'must be a non-empty relative API path', path)
    }
    return new URL(normalized, this.baseUrl)
  }
}

export const HA_LIST_ENTITIES_DEFINITION: ToolDefinition = {
  type: 'function',
  function: {
    name: 'ha_list_entities',
    description: 'List Home Assistant entity states, optionally filtered by domain and area.',
    parameters: {
      type: 'object',
      additionalProperties: false,
      properties: {
        area: { type: 'string', description: 'Optional Home Assistant area ID or area name.' },
        domain: { type: 'string', description: 'Optional entity domain, such as light or switch.' },
        limit: { type: 'integer', minimum: 1, maximum: MAX_ENTITY_LIMIT, default: DEFAULT_ENTITY_LIMIT },
      },
    },
  },
}

export const HA_LIST_SERVICES_DEFINITION: ToolDefinition = {
  type: 'function',
  function: {
    name: 'ha_list_services',
    description: 'List Home Assistant service domains, optionally limited to one domain.',
    parameters: {
      type: 'object',
      additionalProperties: false,
      properties: {
        domain: { type: 'string', description: 'Optional service domain, such as light.' },
      },
    },
  },
}

export const HA_GET_STATE_DEFINITION: ToolDefinition = {
  type: 'function',
  function: {
    name: 'ha_get_state',
    description: 'Get the current state, attributes, and timestamps for one Home Assistant entity.',
    parameters: {
      type: 'object',
      additionalProperties: false,
      properties: {
        entity_id: { type: 'string', description: 'Entity ID, such as light.living_room.' },
      },
      required: ['entity_id'],
    },
  },
}

export const HA_CALL_SERVICE_DEFINITION: ToolDefinition = {
  type: 'function',
  function: {
    name: 'ha_call_service',
    description: 'Call an explicitly configured Home Assistant service.',
    parameters: {
      type: 'object',
      additionalProperties: false,
      properties: {
        data: { type: 'object', description: 'JSON service data, for example an entity_id.' },
        domain: { type: 'string', description: 'Service domain, such as light.' },
        service: { type: 'string', description: 'Service name, such as turn_on.' },
      },
      required: ['domain', 'service'],
    },
  },
}

export const HOME_ASSISTANT_TOOL_DEFINITIONS: readonly ToolDefinition[] = [
  HA_LIST_ENTITIES_DEFINITION,
  HA_LIST_SERVICES_DEFINITION,
  HA_GET_STATE_DEFINITION,
  HA_CALL_SERVICE_DEFINITION,
]

/**
 * Host-bound configuration for Home Assistant tools.
 *
 * A resolver supports multi-session hosts without a mutable process-global
 * client. The static client is useful for single-installation deployments.
 */
export interface HomeAssistantToolsOptions {
  readonly client?: HomeAssistantClient
  readonly resolveClient?: (context: ToolExecutionContext) => HomeAssistantClient | undefined
}

/** Register Home Assistant tools only after the host supplies an explicit client or resolver. */
export function registerHomeAssistantTools(registry: ToolRegistry, options: HomeAssistantToolsOptions): void {
  if (!options.client && !options.resolveClient) {
    throw new ConfigurationError('homeAssistant.client', 'must be supplied when registering Home Assistant tools')
  }

  registry.register(HA_LIST_ENTITIES_DEFINITION, (inputs, context, signal) => (
    listHomeAssistantEntities(inputs, clientFor(options, context), signal)
  ))
  registry.register(HA_LIST_SERVICES_DEFINITION, (inputs, context, signal) => (
    listHomeAssistantServices(inputs, clientFor(options, context), signal)
  ))
  registry.register(HA_GET_STATE_DEFINITION, (inputs, context, signal) => (
    getHomeAssistantState(inputs, clientFor(options, context), signal)
  ))
  registry.register(HA_CALL_SERVICE_DEFINITION, (inputs, context, signal) => (
    callHomeAssistantService(inputs, clientFor(options, context), signal)
  ))
}

/** List and summarize entities in the Python tool's stable JSON shape. */
export async function listHomeAssistantEntities(
  inputs: JsonObject,
  client: HomeAssistantClient,
  signal?: AbortSignal,
): Promise<JsonObject> {
  const domain = optionalDomain(inputs, 'domain')
  const area = optionalArea(inputs, 'area')
  const limit = entityLimit(inputs)

  try {
    const states = await client.listStates(signal)
    const entities = states
      .filter(state => stateMatches(state, domain, area))
      .slice(0, limit)
      .map(stateSummary)
    return { count: entities.length, entities }
  } catch (error) {
    return { count: 0, entities: [], error: toolErrorCode(error) }
  }
}

/** List services in the Python tool's stable JSON shape. */
export async function listHomeAssistantServices(
  inputs: JsonObject,
  client: HomeAssistantClient,
  signal?: AbortSignal,
): Promise<JsonObject> {
  const domain = optionalDomain(inputs, 'domain')
  try {
    const catalog = await client.listServices(signal)
    const domains = domain === undefined
      ? catalog
      : catalog.filter(serviceDomain => serviceDomain.domain === domain)
    return { domains: [...domains] }
  } catch (error) {
    return { domains: [], error: toolErrorCode(error) }
  }
}

/** Return one entity state, with a safe structured result for missing or failed reads. */
export async function getHomeAssistantState(
  inputs: JsonObject,
  client: HomeAssistantClient,
  signal?: AbortSignal,
): Promise<JsonObject> {
  const entityId = homeAssistantEntityId(requiredString(inputs, 'entity_id'))
  try {
    const state = await client.getState(entityId, signal)
    if (!state || typeof state.entity_id !== 'string') {
      return { entity_id: entityId, error: 'not_found' }
    }
    return stateDetail(state)
  } catch (error) {
    return { entity_id: entityId, error: toolErrorCode(error) }
  }
}

/** Call a Home Assistant service and return a body-safe status result. */
export async function callHomeAssistantService(
  inputs: JsonObject,
  client: HomeAssistantClient,
  signal?: AbortSignal,
): Promise<JsonObject> {
  const domain = homeAssistantDomain(requiredString(inputs, 'domain'), 'domain')
  const service = homeAssistantDomain(requiredString(inputs, 'service'), 'service')
  const data = serviceData(inputs)

  try {
    const changed = await client.callService(domain, service, data, signal)
    return { changed: [...changed], domain, ok: true, service }
  } catch (error) {
    return { domain, error: toolErrorCode(error), ok: false, service }
  }
}

/** Filter raw state records by optional Home Assistant domain and area metadata. */
export function filterHomeAssistantEntities(
  states: readonly JsonObject[],
  domain: string | undefined = undefined,
  area: string | undefined = undefined,
): JsonObject[] {
  return states.filter(state => stateMatches(state, domain, area))
}

function clientFor(options: HomeAssistantToolsOptions, context: ToolExecutionContext): HomeAssistantClient {
  const resolved = options.resolveClient?.(context) ?? options.client
  if (!resolved) {
    throw new ConfigurationError('homeAssistant.client', 'is not configured for this tool context')
  }
  return resolved
}

function stateMatches(state: JsonObject, domain: string | undefined, area: string | undefined): boolean {
  const entityId = typeof state.entity_id === 'string' ? state.entity_id : ''
  if (domain !== undefined && !entityId.startsWith(domain + '.')) {
    return false
  }
  if (area === undefined) {
    return true
  }
  const attributes = isJsonObject(state.attributes) ? state.attributes : {}
  return attributes.area_id === area || attributes.area === area
}

function stateSummary(state: JsonObject): JsonObject {
  return {
    attributes: isJsonObject(state.attributes) ? state.attributes : {},
    entity_id: typeof state.entity_id === 'string' ? state.entity_id : null,
    state: typeof state.state === 'string' ? state.state : null,
  }
}

function stateDetail(state: JsonObject): JsonObject {
  return {
    attributes: isJsonObject(state.attributes) ? state.attributes : {},
    entity_id: typeof state.entity_id === 'string' ? state.entity_id : null,
    last_changed: typeof state.last_changed === 'string' ? state.last_changed : '',
    last_updated: typeof state.last_updated === 'string' ? state.last_updated : '',
    state: typeof state.state === 'string' ? state.state : null,
  }
}

function serviceData(inputs: JsonObject): JsonObject {
  const data = inputs.data
  if (data === undefined) {
    return {}
  }
  if (!isJsonObject(data)) {
    throw new ValidationError('data', 'must be a JSON object', data)
  }
  return data
}

function optionalDomain(inputs: JsonObject, field: string): string | undefined {
  const value = optionalString(inputs, field)
  return value === undefined ? undefined : homeAssistantDomain(value, field)
}

function optionalArea(inputs: JsonObject, field: string): string | undefined {
  const value = optionalString(inputs, field)
  if (value === undefined) {
    return undefined
  }
  const normalized = value.trim()
  if (!normalized) {
    throw new ValidationError(field, 'must be a non-empty string when supplied', value)
  }
  return normalized
}

function entityLimit(inputs: JsonObject): number {
  const limit = optionalInteger(inputs, 'limit', DEFAULT_ENTITY_LIMIT)
  if (limit < 1 || limit > MAX_ENTITY_LIMIT) {
    throw new ValidationError('limit', 'must be between 1 and ' + MAX_ENTITY_LIMIT, limit)
  }
  return limit
}

function homeAssistantEntityId(value: string): string {
  const normalized = value.trim()
  if (!ENTITY_ID_PATTERN.test(normalized)) {
    throw new ValidationError('entity_id', 'must be a valid Home Assistant entity ID', value)
  }
  return normalized
}

function homeAssistantDomain(value: string, field: string): string {
  const normalized = value.trim()
  if (!DOMAIN_PATTERN.test(normalized)) {
    throw new ValidationError(field, 'must contain only lowercase letters, digits, and underscores', value)
  }
  return normalized
}

function normalizeBaseUrl(value: string): URL {
  let parsed: URL
  try {
    parsed = new URL(value)
  } catch (error) {
    throw new ConfigurationError('homeAssistant.baseUrl', 'must be an absolute HTTP(S) URL', { cause: error })
  }
  if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') {
    throw new ConfigurationError('homeAssistant.baseUrl', 'must use HTTP or HTTPS')
  }
  if (parsed.username || parsed.password) {
    throw new ConfigurationError('homeAssistant.baseUrl', 'must not embed credentials')
  }
  if (parsed.search || parsed.hash) {
    throw new ConfigurationError('homeAssistant.baseUrl', 'must not include a query string or fragment')
  }
  if (!parsed.pathname.endsWith('/')) {
    parsed.pathname += '/'
  }
  return parsed
}

function normalizedToken(value: string | undefined): string {
  const token = value?.trim() ?? ''
  if (token.includes('\n') || token.includes('\r')) {
    throw new ConfigurationError('homeAssistant.token', 'must not contain line breaks')
  }
  return token
}

function boundedTimeout(value: number): number {
  if (!Number.isInteger(value) || value < 1 || value > MAX_TIMEOUT_MS) {
    throw new ConfigurationError('homeAssistant.timeoutMs', 'must be an integer between 1 and ' + MAX_TIMEOUT_MS)
  }
  return value
}

function requestSignal(signal: AbortSignal | undefined, timeoutMs: number): AbortSignal {
  const timeout = AbortSignal.timeout(timeoutMs)
  return signal === undefined ? timeout : AbortSignal.any([signal, timeout])
}

function requiredObjectArray(value: JsonValue | undefined, field: string): JsonObject[] {
  if (!Array.isArray(value)) {
    throw new HomeAssistantRequestError('Home Assistant returned a malformed ' + field + ' response')
  }
  const objects: JsonObject[] = []
  for (const entry of value) {
    if (!isJsonObject(entry)) {
      throw new HomeAssistantRequestError('Home Assistant returned a malformed ' + field + ' response')
    }
    objects.push(entry)
  }
  return objects
}

function requiredObject(value: JsonValue, field: string): JsonObject {
  if (!isJsonObject(value)) {
    throw new HomeAssistantRequestError('Home Assistant returned a malformed ' + field + ' response')
  }
  return value
}

function isJsonValue(value: unknown): value is JsonValue {
  if (
    value === null
    || typeof value === 'boolean'
    || typeof value === 'number'
    || typeof value === 'string'
  ) {
    return true
  }
  if (Array.isArray(value)) {
    return value.every(isJsonValue)
  }
  return isJsonObject(value) && Object.values(value).every(isJsonValue)
}

function toolErrorCode(error: unknown): string {
  if (error instanceof ConfigurationError) {
    return 'not_configured'
  }
  if (error instanceof HomeAssistantRequestError) {
    if (error.status === 401 || error.status === 403) {
      return 'unauthorized'
    }
    if (error.status === 404) {
      return 'not_found'
    }
  }
  return 'request_failed'
}
